virtual_ship/make_realistic/ctd_make_realistic.py

```diff
@@ -0,0 +1,187 @@
+"""ctd_make_realistic function."""
+
+import random
+
+import numpy as np
+import opensimplex
+import py
+import xarray as xr
+
+
+def ctd_make_realistic(
+    zarr_path: str | py.path.LocalPath, out_dir: str | py.path.LocalPath, prefix: str
+) -> list[py.path.LocalPath]:
+    """
+    Take simulated CTD data, add noise, then save in CNV format (1 file per CTD).
+
+    :param zarr_path: Input simulated data.
+    :param out_dir: Output directory for CNV file.
+    :param prefix: Prefix for CNV files. Will be postfixed with '_{ctd_num}'.
+    :returns: Paths to created file.
+    """
+    original = xr.open_zarr(zarr_path)
+
+    files = []
+    for ctd_i, traj in enumerate(original.trajectory):
+        time = original.sel(trajectory=traj)["time"].values
+        reltimesec = (time - time[0]) / np.timedelta64(1, "s")
+        latitude = original.sel(trajectory=traj)["lat"].values
+        longitude = original.sel(trajectory=traj)["lon"].values
+        temperature = original.sel(trajectory=traj)["temperature"].values
+        salinity = original.sel(trajectory=traj)["salinity"].values
+        depth = original.sel(trajectory=traj)["z"].values
+
+        temperature = _add_temperature_noise(temperature, depth)
+        salinity = _add_salinity_noise(salinity, depth)
+
+        out_file = (
+            out_dir.join(f"{prefix}{ctd_i}.cnv")
+            if isinstance(out_dir, py.path.LocalPath)
+            else f"{out_dir}/{prefix}{ctd_i}.cnv"
+        )
+        files.append(out_file)
+
+        cnv = _to_cnv(
+            filename=str(out_file),
+            latitudes=latitude,
+            longitudes=longitude,
+            times=reltimesec,
+            temperatures=temperature,
+            depths=depth,
+            salinities=salinity,
+        )
+
+        with open(out_file, "w") as out_cnv:
+            out_cnv.write(cnv)
+
+    return files
+
+
+def _add_temperature_noise(temperature: np.ndarray, depth: np.ndarray) -> np.ndarray:
+    surface_noise = (
+        2.0
+        * (np.random.random_sample(len(temperature)) * 2.0 - 1)
+        * np.exp(depth / 4.0)
+    )
+    noise = (
+        0.05
+        * np.maximum(1.0 + depth / 1000.0, 0.1)
+        * np.random.random_sample(len(temperature))
+    )
+    drift = (
+        0.3
+        * (np.random.random() * 2.0 - 1)
+        * np.array(range(len(temperature)))
+        / len(temperature)
+    )
+    opensimplex.seed(np.random.randint(999999))
```

Comment:
Why is the seed deteministic? For reproducibility, I guess? But would it not be more realistic to have noise with a random (clock-based?) seed?

Reply:
It is not deterministic :) I seed with a random number created by the random module. The opensimplex library is a bit annoying so I chose to seed it like this. I think we will use something different once we revisit the noise, but for now I think this is fine.